### ISSUE TITLE
i_CTRL-E/i_CTRL-Y and virtual text

### ISSUE THREAD

**Issue by habamax** [vim/vim#11785]
### Steps to reproduce

1. `vim -Nu NONE`
2. Add following lines
```
vim9script

# hello world
prop_type_add('VTError', {highlight: 'Error'})
prop_add(3, 10, {text: '@@@@@@', type: 'VTError'})
```
3. `:syn on`
4. `:source`
5. Goto line 3, `O` and then multiple `CTRL-Y` to add characters

[![asciicast](https://asciinema.org/a/WiM5aU93WsikJtlehImgp9q2v.svg)](https://asciinema.org/a/WiM5aU93WsikJtlehImgp9q2v)

### Expected behaviour

I would expect i_CTRL-Y would take offset of the virtual text above and insert proper char.

### Version of Vim

9.0.1151

### Environment

- WSL Fedora 37
- bash

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> 1. `vim -Nu NONE`
> 2. Add following lines
> ```
> vim9script
> 
> # hello world
> prop_type_add('VTError', {highlight: 'Error'})
> prop_add(3, 10, {text: '@@@@@@', type: 'VTError'})
> ```
> 3. `:syn on`
> 4. `:source`
> 5. Goto line 3, `O` and then multiple `CTRL-Y` to add characters
> 
> [![asciicast](https://asciinema.org/a/WiM5aU93WsikJtlehImgp9q2v.svg)](https://asciinema.org/a/WiM5aU93WsikJtlehImgp9q2v)
> 
> 
> 
> ### Expected behaviour
> 
> I would expect i_CTRL-Y would take offset of the virtual text above
> and insert proper char.

Do you mean it should behave as if the virtual text isn't there?  But
then wat if you start after the virtual text, don't you expect the text
just above/below the cursor to be inserted?  If we ignore virtual text
it would get characters some number of characters to the left.

I'm not sure how to define what should happen here.

-- 
If you don't get everything you want, think of
everything you didn't get and don't want.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by habamax** [comment]
> Do you mean it should behave as if the virtual text isn't there? 

Yes, and take chars after virtual text considering real char position.

If this is too cumbersome, then maybe document current behaviour. document

**Comment by habamax** [comment]
My real case was with a simple one char virtual text, which represents color:
<a href="https://asciinema.org/a/fSqMbrr0mvyWixIdyhCTJX35O" target="_blank"><img src="https://asciinema.org/a/fSqMbrr0mvyWixIdyhCTJX35O.svg" /></a>

**Comment by bfrg** [comment]
IMO the current behavior looks like a bug. Since `yy` and `getline()` return only the actual buffer text, I would also expect `i_CTRL-Y` to behave similarly.

**Comment by brammool** [comment]
> IMO the current behavior looks like a bug. Since `yy` and `getline()`
> return only the actual buffer text, I would also expect `i_CTRL-Y` to
> behave similarly.

Well, it does.  It fetches the character in the line above or below at
the same screen column.  It skips over the virtual text, since that
isn't actual buffer text.

Like I said, any another way of doing this is not going to be better.
When not using the virtual text for computing the position then the
column will be wrong whenever there is virtual text before the current
position.

Another way would be to consider the virtual text as a gap, and always
use a space character.  I don't see how that is more useful, but it
might look better.

-- 
OLD WOMAN: King of the WHO?
ARTHUR:    The Britons.
OLD WOMAN: Who are the Britons?
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by bfrg** [comment]
In the above example we have the following line: `# hello w@@@@@@orld`, where `@@@@@@` is virtual text.

Yanking and pasting the entire line, I end up with:
```
# hello w@@@@@@orld
# hello world
```

But when I press `i_CTRL-Y`, I end up with:
```
# hello w@@@@@@orld
# hello wooooooorld
```

That is wrong in my opinion. I would have expected:
```
# hello w@@@@@@orld
# hello world
```

Adding space characters wouldn't look right either IMHO.

**Comment by brammool** [comment]
> In the above example we have the following line: `# hello w@@@@@@orld`, where `@@@@@@` is virtual text.
> 
> Yanking and pasting the entire line, I end up with:
> ```
> # hello w@@@@@@orld
> # hello world
> ```
> 
> But when I press `i_CTRL-Y`, I end up with:
> ```
> # hello w@@@@@@orld
> # hello wooooooorld
> ```
> 
> That is wrong in my opinion. I would have expected:
> ```
> # hello w@@@@@@orld
> # hello world
> ```
> 
> Adding space characters wouldn't look right either IMHO.

This might work if we count the number of characters of the virtual text
that we already skipped.  E.g., when below the third "@" we have seen
two "@" already, thus for the trailing "orld" we skip two characters and
get the "l".

What if you start using CTRL-Y halfway the virtual text:

```
# hello w@@@@@@orld
# some text
```

Now if we apply the same rules we get:


```
# hello w@@@@@@orld
# some textld
```

That doesn't make any sense, does it?

And what if we are further down the line:

```
# hello w@@@@@@orld with some more text here
# some text sdf asdf asdf asd
```

When using CTRL-Y here, you probably want to copy the word "more":

```
# hello w@@@@@@orld with some more text here
# some text sdf asdf asdf asd more text here
```

However, if we would take the virtual text into account you would get:

```
# hello w@@@@@@orld with some more text here
# some text sdf asdf asdf asd ext here
```

Not very useful either.

-- 
BLACK KNIGHT: I'm invincible!
ARTHUR:       You're a looney.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dylanahsmith** [comment]
In my opinion, the desired user behaviour when using `i_CTRL-Y` below virtual text would be to temporarily insert virtual text padding as well as the character after the virtual text, then to remove that virtual text padding when leaving insert mode.

For example, when trying to copy the above characters with `i_CTRL-Y` in the original given example, you would
end up with the following in insert mode after pressing CTRL-Y 4 times, with `@̶@̶@̶@̶@̶@̵` being used to represent the temporary virtual text padding while in insert mode

```
# hello w@@@@@@orld
# hello w@̶@̶@̶@̶@̶@̵orld
```

Leaving insert mode would remove the temporary virtual text padding, leaving the text as the following by not copying the above virtual text into the buffer content itself

```
# hello w@@@@@@orld
# hello world
```

That way it would intuitively copy the above character, even if it were after the virtual text.  For example, in Bram's example, copying the above character from the following

```
# hello w@@@@@@Orld with some more text here
# some text sdf asdf asdf asd
```

would intuitively copy ` more`.

As I've tried to illustrate with Unicode combining characters, the temporary virtual text could use the strikethrough attribute in order to communicate that it is temporary, but can otherwise re-use the same characters and attributes as the above line to help communicate that it is still virtual text.  Also, having the first use of `i_CTRL-Y` below virtual text both add the temporary virtual text and copy a single character from the buffer contents should make it clearer that it has added virtual text, similar to how moving the cursor over virtual text skips over it.

Thoughts on whether that seems like the desired behaviour?

From an implementation perspective, using virtual text to achieve this behaviour seems like it would just require:
* Hooking into the existing check to see if the virtual text is above the cursor when `i_CTRL-E`/`i_CTRL-Y` to add the virtual text.
* Using a hook like `InsertLeave` to remove that virtual text.

**Comment by chrisbra** [comment]
but that seems odd, if the text retrieved via `CTRL-Y` suddenly disappears.
